internal/smartlink/smartlink.go

```diff
@@ -80,6 +90,32 @@ func Link(src, dest string) error {
 	return nil
 }
 
+func isCircularLink(src string) (bool, error) {
+	if !fileutils.IsSymlink(src) || !fileutils.IsDir(src) {
+		return false, nil
+	}
+
+	target, err := os.Readlink(src)
+	if err != nil {
+		return false, errs.Wrap(err, "could not read symlink %s", src)
+	}
+
+	if !filepath.IsAbs(target) {
+		resolved, err := fileutils.ResolveUniquePath(src)
+		if err != nil {
+			return false, errs.Wrap(err, "Could not resolve src path")
+		}
+		target = resolved
+	}
+
+	// If the target points at the parent directory, we have a circular link
+	if filepath.Dir(src) == target {
+		return true, nil
+	}
+
+	return false, nil
+}
```

Comment:
This doesn't feel right. It's totally valid to have a link like this:

```
- dir1/file1
- dir1/dir2 > ../dir1
```

The problem is that our Linking mechanism doesn't understand this and will continue to recurse forever. We need to fix that, whilst still supporting a link that goes back up the path.

Reply:
Since this library doesn't link directories I'm not sure how we can support this beyond what this PR is doing. If we encounter a circular directory we skip it, all of the contents of the directory will still be linked.

 I've moved the check to make it more clear. I think that in the end, all we can do is ensure all of the files in the directories and subdirectories have been linked with the same layout as the source.